tests for creating glance image

jsuchome opened the pull request:
Covers tests 107-109 from https://github.com/SUSE-Cloud/qa-openstack-cli/tree/master/tests

Commit:
checking for remote file presence

Commit:
added tests for glance image creating

config file for glance testing

base glance testing only on new client.openstack commands

aspiers commented on features/glance.feature:
```diff
@@ -0,0 +1,39 @@
+@glance
```
Looking at this file makes me very happy :-)

aspiers commented on features/glance.feature:
```diff
@@ -0,0 +1,39 @@
+    Then the image is listed as active
+    And this image has non-empty ID
+    And its ID can be used to show the image info
+    When I delete the image
```
I think deletion should be in a separate scenario.

jsuchome replied:
Should it? I wanted to do some cleanup after the scenario is done, so nothing created for testing is left there. So I actually think it is better to delete it right after the test finishes. @vmoravec what do you think?

aspiers replied:
Cleanup (i.e. test teardown) is different from actual testing.  Currently you are including deletion as part of the actual test.  Whilst I think it is right to properly test deletion, it should not be mixed in with testing of creation, because then if the test fails, you have to spend extra effort to determine whether it failed due to creation or deletion - it's basically confusing things by mixing multiple things into a single scenario.

In contrast, if you only want to delete as part of teardown, then it should be done outside the body of the test, e.g. via [an `After` hook for the scenario](https://github.com/cucumber/cucumber/wiki/Hooks#global-hooks).

jsuchome replied:
I've thought about After hooks, but I do not see how could I limit it to just one scenario. Or, after each scenario for just one given feature....

aspiers replied:
Just use a `@cleanup_image` tag.

vmoravec replied:
@aspiers I don't agree that removing of an image is different from testing. Everything what is being done should be captured in the scenario text. I consider that as best practice, otherwise we end up with a lot of action hidden before the our eyes when reading the feature and spenario specifications.

It's also not practical to put that into an another scenario because every scenario must be independent, in this case we would create dependency by calling one scenario from the other. In case the latter fails, you end up looking for the reason why this particular scenario was run. It is clear e.g. for a `Background`, not for other cases.

I would not put constraints on what is actual testing and what not. If the scenario fails while the image is being deleted, it's straight forward to see it, more effort is actually not needed compared to a separate scenario having failed, the reason will be the same.

Additionally, I would like to avoid using hooks as it could end up messy,  and rather keep the scenarios' as simple as possible and build mainly on the expressivnes of Gherkin language.

aspiers replied:
> @aspiers I don't agree that removing of an image is different from testing.

I didn't say it was - I said _cleanup_ is different from actual testing.

> Everything what is being done should be captured in the scenario text.

But that's not really true - there is a TON of setup done which is not captured by the scenario text (setup of the nodes, installation of SUSE Cloud, creation and application of Crowbar barclamp proposals...).

> I consider that as best practice, otherwise we end up with a lot of action hidden before the our eyes when reading the feature and spenario specifications.

As per above we already have a lot of stuff hidden.  But that's fine.  The scenario is only supposed to test one thing.

> It's also not practical to put that into an another scenario because every scenario must be independent, in this case we would create dependency by calling one scenario from the other.  In case the latter fails, you end up looking for the reason why this particular scenario was run. It is clear e.g. for a Background, not for other cases.

No, I think you misunderstand me.  I am definitely not suggesting to create a dependency (and even if I was, I'm not sure if Cucumber actually supports doing that).

> I would not put constraints on what is actual testing and what not. If the scenario fails while the image is being deleted, it's straight forward to see it, more effort is actually not needed compared to a separate scenario having failed, the reason will be the same.

I disagree - the problem is that it breaks the test suite report, because if something breaks with image deletion, any scenario which has image deletion at the end will be reported as failing.  That's bad, because a) it makes the situation look worse than it actually is, and b) extra effort is required to check all failing scenarios and ascertain that they're all only due to a single issue with image deletion.

In contrast, if image deletion is moved to a teardown hook, and tested explicitly via an independent image deletion scenario, then if it fails, you only get a single failure in the report, and you instantly know without any further examination that the failure is directly related to image deletion.

> Additionally, I would like to avoid using hooks as it could end up messy

Why?

> and rather keep the scenarios' as simple as possible

Yes I agree - my proposal makes them simpler than they currently are, because it avoids polluting scenarios with extra steps for teardown.  For example look at http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html - the declarative style is much more readable mainly because the scenario is much smaller.

vmoravec replied:
> Everything what is being done should be captured in the scenario text.
> 
> > But that's not really true - there is a TON of setup done which is not captured by the scenario text (setup of the >> nodes, installation of SUSE Cloud, creation and application of Crowbar barclamp proposals...).  

Yes, setup and teardown of the environment in broader sense is not done inside the cucumber tests, here we need to synchronize with the deployment paths. This is really not captured by the scenarios and was not meant to. 

> I consider that as best practice, otherwise we end up with a lot of action hidden before the our eyes when reading the feature and spenario specifications.
> 
> > As per above we already have a lot of stuff hidden.  But that's fine.  The scenario is only supposed to test one thing.

Cucumber tests are usually self-driven from inside and capture everything in step definitions. Unfortunately, with our approach we will always have two contexts - `mkcloud` infrastructure setup and "dumb" `cucumber` tests where the knowledge about the current setup is mandatory for the person who runs the tests. But this does not limit us for how to build the scenarios.

> I would not put constraints on what is actual testing and what not. If the scenario fails while the image is being deleted, it's straight forward to see it, more effort is actually not needed compared to a separate scenario having failed, the reason will be the same.
> 
> > I disagree - the problem is that it breaks the test suite report, because if something breaks with image deletion, any scenario which has image deletion at the end will be reported as failing.  That's bad, because a) it makes the situation look worse than it actually is, and b) extra effort is required to check all failing scenarios and ascertain that they're all only due to a single issue with image deletion.

Maybe we both mean different things. I don't see the action of deletion as a generic event that is equal with all scenarios you need to check for. 

I went to [cucumber mailing list](https://groups.google.com/forum/#!forum/cukes) to get some information from those who designed this tool and found out that I was wrong with suggesting explicit image deletion at the end of a scenario. The recommended way here is to do the cleanup in the `Given` steps. It's important to add here, that we don't do any generic cleanup, we just need to get rid of a specific file if it exists. This is not (easily) doable in a separate scenario or hook due to scenario specific variables/config.

> Additionally, I would like to avoid using hooks as it could end up messy
> 
> > Why?

Every scenario would expect new hook definition. As these hooks are defined in separate files in `features/support/some_file.rb` we would need to keep that always with sync with the scenarios and features, give these hooks some names, maybe create categories etc; but this is not the main issue here. The problem is with the specific config and variables for the scenarios - the hooks apply mostly for general things that are not dependent on any things defined within a scenario. Additionally hooks add more effort for newcomers to the testsuite to grasp it easily.

> and rather keep the scenarios' as simple as possible
> 
> > Yes I agree - my proposal makes them simpler than they currently are, because it avoids polluting scenarios with extra steps for teardown.  For example look at http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html - the declarative style is much more readable mainly because the scenario is much smaller.

As I mentioned above, if we cover such cases in the `Given` steps (like `Given the image "jeos1" does not exist`), we keep the scenarios clean.

aspiers replied:
> Cucumber tests are usually self-driven from inside and capture everything in step definitions. Unfortunately, with our approach we will always have two contexts - `mkcloud` infrastructure setup and "dumb" `cucumber` tests where the knowledge about the current setup is mandatory for the person who runs the tests. But this does not limit us for how to build the scenarios.

Agreed.

> I would not put constraints on what is actual testing and what not. If the scenario fails while the image is being deleted, it's straight forward to see it, more effort is actually not needed compared to a separate scenario having failed, the reason will be the same.
> 
> > I disagree - the problem is that it breaks the test suite report, because if something breaks with image deletion, any scenario which has image deletion at the end will be reported as failing.  That's bad, because a) it makes the situation look worse than it actually is, and b) extra effort is required to check all failing scenarios and ascertain that they're all only due to a single issue with image deletion.
> 
> Maybe we both mean different things.

Maybe ;-)

> I don't see the action of deletion as a generic event that is equal with all scenarios you need to check for.

Agreed.  Sometimes deletion is the thing you are focussing on testing, and other times it is just part of setup and/or teardown.

> I went to [cucumber mailing list](https://groups.google.com/forum/#!forum/cukes) to get some information from those who designed this tool and found out that I was wrong with suggesting explicit image deletion at the end of a scenario. The recommended way here is to do the cleanup in the `Given` steps.

That sounds like a good approach, yes.

> It's important to add here, that we don't do any generic cleanup, we just need to get rid of a specific file if it exists.

That's not necessarily sufficient.  For example if you had a bunch of scenarios which all created new images with different names, each `Given the image "$NAME" does not exist` would be different because `$NAME` would be different from each.  So then you would also need each scenario to have another step `Given there is enough unused quota for image "$NAME"`, otherwise there would be a risk of running out of image quota due to lack of image deletion at the end of each scenario.

Yes this is a slightly artificial example, but it would become a lot more real if we were talking about guest VM instances in nova rather than images in glance.

However the good news is that this could be done very nicely via [backgrounds](https://github.com/cucumber/cucumber/wiki/Background)!

> This is not (easily) doable in a separate scenario or hook due to scenario specific variables/config.

I don't see why not - sounds easily doable to me using a hook.

> Additionally, I would like to avoid using hooks as it could end up messy
> 
> > Why?
> 
> Every scenario would expect new hook definition.

Really?  Couldn't you simply reuse the same image cleanup hook across all scenarios which create images?

```
After('@creates_kvm_image') do
  # Delete the image which was created
  control_node.openstack.image.delete @image_id
end
```

> As these hooks are defined in separate files in `features/support/some_file.rb` we would need to keep that always with sync with the scenarios and features

Sure, but that is already true of all steps.  That's just the way Cucumber works.

> give these hooks some names, maybe create categories etc;

I don't think hooks need names or categories.  It's enough to associate them with tags.

> The problem is with the specific config and variables for the scenarios - the hooks apply mostly for general things that are not dependent on any things defined within a scenario.

I don't see the problem - doesn't the use of @image_id in the above hook take care of that?

> Additionally hooks add more effort for newcomers to the testsuite to grasp it easily.

Typically newcomers will care more about the body of the test rather than setup or teardown.

> and rather keep the scenarios' as simple as possible
> 
> > Yes I agree - my proposal makes them simpler than they currently are, because it avoids polluting scenarios with extra steps for teardown.  For example look at http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html - the declarative style is much more readable mainly because the scenario is much smaller.
> 
> As I mentioned above, if we cover such cases in the `Given` steps (like `Given the image "jeos1" does not exist`), we keep the scenarios clean.

Yes, I like that approach a lot.  As I already said, you may also need to check that quotas are not close to being exceeded, and it would make sense to consider [backgrounds](https://github.com/cucumber/cucumber/wiki/Background) for achieving this.

Thanks for the really good discussion!

aspiers commented on features/glance.feature:
```diff
@@ -0,0 +1,39 @@
+    Then the image is listed as active
+    And this image has non-empty ID
+    And its ID can be used to show the image info
+    When I delete the image
```
Do we need to test deletion for all the different image types, or maybe is one enough?

aspiers commented on features/glance.feature:
```diff
@@ -0,0 +1,39 @@
+  As an administrator
+  I want to make sure that Glance component is working correctly
+
+  @create_kvm_image
```
Just curious what the point of this tag is?  I didn't think tags were intended to be used for labelling individual scenarios, but more for tagging scenarios which have a particular characteristic, e.g. I would suggest that this scenario has the `@kvm` tag, and maybe `@slow`.

jsuchome replied:
If nothing else, it is referenced from config/features/glance.yml

aspiers replied:
OK, so I added a review comment to `config/features/glance.yml` - I think this is the wrong choice of tag name.

vmoravec replied:
@aspiers Making use of unique tags for scenarios is important for calling the particular scenarios from rake tasks. Yes, we end up with a tag for every single scenario, but that's the price. On the other hand, we are making use of it for feature specific configuration and rake tasks so far. Still it does not limit us to use other scenario tags in different contexts.

aspiers replied:
I see.  Why do we need a rake task for each scenario though?  That doesn't sound like it will scale very well.

vmoravec replied:
It's for convenience, it's useful when writing and testing a scenario and it's useful when you want to list scenarios for a specific feature with `rake` command (no need to look for the feature file and read the text). 

This is easy to drop when they prove to cause more overhead than benefit later.

aspiers replied:
OK thanks for the info.  I don't know enough about `cct` yet to offer anything concrete yet, but I suspect there may be a way to do this in a more general fashion which scales easily by not requiring manual tagging of each scenario.  For example, http://stackoverflow.com/questions/825748/how-do-i-pass-command-line-arguments-to-a-rake-task

vmoravec replied:
Yep, that's the only alternative I think. I will adapt the code to accept an env var for a feature `rake` task.

aspiers commented on config/features/glance.yml:
```diff
@@ -0,0 +1,15 @@
+features:
+  glance:
+    flags: ""
+    create_kvm_image:
```
Wouldn't it be better to call this `kvm_test_image`?  By de-coupling the name from the scenario, it will allow reusing of the image for other scenarios.

jsuchome replied:
But I'm using it intentionally for the scenario, so I can reference it this way:
https://github.com/jsuchome/cct/blob/glance_to_master/features/step_definitions/glance/image_create_steps.rb#L4

jsuchome replied:
But yes, it might be usable for other scenarios as well. I think potential solution will become more clear once we start working on such scenarios.

aspiers replied:
It would be nicer to avoid the coupling from the beginning, so that if other people extend these features with other scenarios which need the same config, they can immediately reuse it rather than giving in to the temptation to copy'n'paste the config.  But if we both agree to watch out for DRY violations then I guess we can tolerate this coupling for now :)

vmoravec replied:
The coupling is actually intenational. If someone wants to reuse the config values within the same feature, he must create a new section in config. Within a config this might be kept dry by using yaml anchors and references.

vmoravec replied:
Like this:

``` yaml
features:
  glance:
    flags: ""
    create_kvm_image: &kvm_image
      image_name: jeos1
      image_source: http://clouddata.cloud.suse.de/images/jeos-64.qcow2
      properties: "--property hypervisor_type=kvm"
    scenario_nr_2:
    <<: *kvm_image
```

aspiers replied:
That looks the wrong way round to me, because you're still coupling generic KVM setup config to a specific scenario, and then coupling other scenarios to that.  So for example if you later wanted to remove `create_kvm_image` then it's a mess because you would have to relocate that config to somewhere else.  If you wanted to do it using YAML anchors and refs then something like this would be cleaner:

``` yaml
features:
  glance:
    flags: ""
    kvm_image: &kvm_image
      image_name: jeos1
      image_source: http://clouddata.cloud.suse.de/images/jeos-64.qcow2
      properties: "--property hypervisor_type=kvm"
    create_kvm_image:
    <<: *kvm_image
    scenario_nr_2:
    <<: *kvm_image
```

vmoravec replied:
Agree, that's much better approach to structure the yaml file here.

vmoravec commented:
As we decided to go with `openstack-client` for testing the functionality, I added (sofar) methods for `image` subcommand:

``` ruby
# returns list of objects with #id and #name properties
control_node.openstack.image.list

# returns image object with all properties as strings
# accepts id or name of the image
control_node.openstack.image.show('5658957e-f0cf-437c-be2e-4b3cc68a1bd8')

# creates an image, accepts name and params as hash
control_node.openstack.image.create "jeos44", disk_format: "qcow2"

# deletes an image
control_node.openstack.image.delete "id_or_name"
```

Additionally, I added a helper method for long lasting operations:

``` ruby
wait_for "Image upload", max: "5 minutes", sleep: "10 seconds" do
  # do the image upload here
  # check if it's uploaded and break if ready
end
```

If the time of the action exceeds the `max` parameter, it will raise a timeout error with the description of the event `Image upload`, so it is clear what caused the timeout.

jsuchome commented:
@vmoravec updated to use your new methods

vmoravec commented on features/step_definitions/glance/image_create_steps.rb:
```diff
@@ -0,0 +1,70 @@
+end
+
+When(/^I create new glance image based on jeos1$/) do
+  control_node.openstack.image.create @image_name, :properties => @properties, :disk_format => "qcow2", :public => true, :container_format => "bare", :copy_from => @image_source
```
Please make this line shorter to fit into github viewport, 84 characters should be ok.

vmoravec commented:
One note to feature naming: I have gone through some docs and mailing list posts regarding cucumber and found that feature names created from software component names are a big no-no. For this reason I would propose to rename this feature from `Glance` to something else, like `Image management` or  'Images' or something similar.  The point here is that the name must describe the feature, it's purpose.

Commit:
rename glance feature to images (to describe the function, not the tool)

move preparation to Background
rename tests to include deleting as a task

vmoravec commented:
Tested on c13. :+1:

aspiers commented:
> For this reason I would propose to rename this feature from Glance to something else, like Image management or 'Images' or something similar.

Luckily the OpenStack community already had the same idea and carried it out :-)  The "real" name for Glance is the OpenStack Image service, e.g. http://docs.openstack.org/admin-guide-cloud/content/section_image-mgmt.html

aspiers commented on features/images.feature:
```diff
@@ -0,0 +1,40 @@
+    Then this image has non-empty ID
+    And its ID can be used to show the image info
+    And the status of the image is active
+    When I delete the image
```
@jsuchome Based on [my discussion with @vmoravec](https://github.com/SUSE-Cloud/cct/pull/2#discussion_r29770123), we are proposing that for each of these 3 scenarios, you can delete the `When I delete the image` / `Then it is no longer listed` steps since they are already covered by `And glance image does not exist`.  But since this is the same for all three scenarios, you could extract it out into a [background](https://github.com/cucumber/cucumber/wiki/Background), e.g.:

``` gherkin
Background:
  Given glance image does not exist
  And image quota has at least 2GB remaining
```

Of course you still need to test that an image is no longer listed after deletion, but this should be done via a dedicated scenario:

``` gherkin
Scenario: Deletion of a KVM image
  Given a glance image based on jeos
  When I delete the image
  Then it is no longer listed
```

jsuchome replied:
I'm not sure. All scenarios are creating different images, so I'm not sure why should I have one scenario for deleting all of them.

And also, I like it more when the situation is back to original state after executing each scenario, not just after running full feature.

Background is nice, yes, e.g. for testing something that is really shared (and for 2 creating scenarios, they have different image id and name) and I wouldn't expect it to delete the image if it exist

aspiers replied:
> I'm not sure. All scenarios are creating different images, so I'm not sure why should I have one scenario for deleting all of them.

You shouldn't.  Noone's proposing that.

> And also, I like it more when the situation is back to original state after executing each scenario, not just after running full feature.

Yes, that's what I'm proposing.  But my suggestion avoids polluting the image _creation_ test with that cleanup.

> Background is nice, yes, e.g. for testing something that is really shared

No, background is not for shared tests, it's for shared setup (context).  As I already showed in the discussion linked above, image cleanup can be part of the setup. 

> (and for 2 creating scenarios, they have different image id and name)

Sure, and as I already said, this is easily handled via `@image_id` and `@image_name`.

> and I wouldn't expect it to delete the image if it exist

If the background has a step `Given glance image does not exist` then why wouldn't you expect it to delete any existing glance image?

Sorry, but it sounds like you still don't understand what I'm proposing.  Maybe reading the linked discussion will help?

jsuchome replied:
I think I do.

You are proposing that Background ensures that the image does not exist yet. That makes some sense, yes, but it doesn't mean that the state of the system is same after the tests are run, because the images are not deleted after the creation.

Unless we are using After hooks, yes, where we can delete them. But the discussion you are pointing to surely doesn't show any agreement if we should do it in the After hooks or not.

In general, I still find current approach more readable and understandable, although yours might be more pure and principally better.

aspiers replied:
> But the discussion you are pointing to surely doesn't show any agreement if we should do it in the After hooks or not.

We didn't reach agreement yet, but yes I am proposing to use both backgrounds and an `After` hook.

> In general, I still find current approach more readable and understandable, although yours might be more pure and principally better.

Well some part of it is personal taste, yes.  Personally I find the combination of this with an `After` hook for cleanup more readable and understandable:

``` gherkin
Background:
  Given glance image does not exist
  And image quota has at least 2GB remaining

Scenario: Creation of XEN HVM image
  Given XEN HVM source image exists
  When I create new glance image based on jeos
  Then this image has non-empty ID
  And its ID can be used to show the image info
  And the status of the image is active
```

because:
- it does not violate the DRY rule by repeating the same 3 steps in every scenario
- each scenario has 5 steps instead of 8, so is easier to read
- each scenario just tests a single thing
- setup and teardown are cleanly separated from the actual test
- the test harness can avoid spamming the report with setup and teardown (which you should never have to care about unless it fails)
- the test harness can more clearly distinguish between failures in tests and failures in setup/teardown
- even if previous teardown failed (e.g. due to a bug in image deletion), it will check that there is still sufficient quota to create the new image

I'm not sure why you would prefer to continue breaking the DRY rule by not using backgrounds or an `After` hook, but even if you do, it would be better to define a nested step e.g.

``` ruby
And /^the image can be deleted$/ do
  steps %{
    When I delete the image
    Then it is no longer listed
  }
end
```

so that at least you are only duplicating 2 lines per scenario instead of 3.

BTW this long conversation about this one PR may feel like pointless bike-shedding, but I think it's actually really worthwhile because `cct` is in its very early days, so any feature-writing style which is established now will certainly be copied by other people as they join in the effort to write new features and steps.  So I think we should try to set a really good precedent for others to follow.

aspiers replied:
Actually the DRY violation is a bit worse than I described, because these 3 steps are also duplicated in every scenario:

``` gherkin
Then this image has non-empty ID
  And its ID can be used to show the image info
  And the status of the image is active
```

So it should be possible to collapse each scenario from 8 steps down to 2 or 3 by using appropriate nested steps.

One more point - these are mostly not my original ideas.  Lots of people advocate DRY in Cucumber features, e.g.
- http://eggsonbread.com/2010/09/06/my-cucumber-best-practices-and-tips/
- http://makandracards.com/makandra/18905-how-to-not-repeat-yourself-in-cucumber-scenarios
- http://rails-bestpractices.com/tag/cucumber/
- http://coryschires.com/ten-tips-for-writing-better-cucumber-steps/

vmoravec commented:
I consider the step `And /^the image can be deleted$/ do` superior to having two steps for the image deletion, but I don't see it as an issue having the step there at all. I think it's good to have and see this action in the scenario text. 

I will repeat myself here:  I would like to avoid using hooks as I only see them as helpers for scenario runners, not as place for implementation of  test specific stuff.

jsuchome commented:
@aspiers OK, so I've created Background and for deleting, I did something different; renamed the tests so they explicitly state that deleting is part of the test (thus it is no teardown). Do you find it wrong?

aspiers commented:
@vmoravec wrote:

> I consider the step And /^the image can be deleted$/ do superior to having two steps for the image deletion, but I don't see it as an issue having the step there at all. I think it's good to have and see this action in the scenario text.

OK, but I still didn't hear an explanation for why you think that.  I gave some concrete reasons why mixing tests and teardown is bad (repeated below).

> I will repeat myself here: I would like to avoid using hooks as I only see them as helpers for scenario runners, not as place for implementation of test specific stuff.

I don't understand this reasoning.  Why would Cucumber support associating hooks with tags if they were not intended for implementation of test-specific stuff?

@jsuchome wrote:

> @aspiers OK, so I've created Background and for deleting, I did something different; renamed the tests so they explicitly state that deleting is part of the test (thus it is no teardown). Do you find it wrong?

Thanks.  This is better than before, for sure.  It still has many of the issues I previously listed though:
- it still violates the DRY rule by repeating the same 4 steps in every scenario
- each scenario tests more than one thing
- teardown is not cleanly separated from the actual test
- the test harness can't avoid spamming the report with teardown (which you should never have to care about unless it fails)
- the test harness can't distinguish between failures in tests and failures in teardown

Thinking about it further, these three scenarios are so close to being identical that I think a [scenario outline](https://github.com/cucumber/cucumber/wiki/Scenario-Outlines) would probably make much more sense.

In summary:
- I'm SUPER happy that you are working on this.
- It's _very_ close to being awesome.
- I know I am being annoying and awkward, but I'm deliberately pushing for the quality of the feature to be (in my eyes) as high as possible, because I expect it will get used as a template for many other features written by other people in the future.  For that reason, and the above reasons, I don't want to give a +1 yet.
- You could ignore me and get two +1s from other people, but I hope you prefer to continue addressing my concerns ;-)

Thanks!

jsuchome commented:
I'm gonna look at that Scenario Outlines, yes (maybe it will solve the rest). For your other concerns:

I'm not too concerned about violating DRY, because the code is not repeated, just the steps descriptions. And I want those steps to be executed for each scenario.

When the tests are defined that they test creating AND deleting, I do not consider deleting as teardown (but yes, than each scenario tests more than one thing)

jsuchome commented:
Hm. So, this is how could it look like with Scenario Outline: http://pastebin.suse.de/13809. Output looks like http://pastebin.suse.de/13810/src

Looks nice, but there's no use for the Background now, because for every example value I have to explicitly look into the config file for config values. So I do not know in Background for which image name to look (I could also make them part of Example, but than the'd pollute the test/step names which is not nice)

vmoravec commented:
@aspiers 

> I consider the step And /^the image can be deleted$/ do superior to having two steps for the image deletion, but I don't see it as an issue having the step there at all. I think it's good to have and see this action in the scenario text.
> 
> > OK, but I still didn't hear an explanation for why you think that.  I gave some concrete reasons why mixing tests and teardown is bad (repeated below).

One reason is that the `delete` action is not a teardown, but rather part of the test, it belongs to the images' management features and it's actually good to have it there explicitly because everyone can see it's being tested. 

> I will repeat myself here: I would like to avoid using hooks as I only see them as helpers for scenario runners, not as place for implementation of test specific stuff.
> 
> > I don't understand this reasoning.  Why would Cucumber support associating hooks with tags if they were not intended for implementation of test-specific stuff?

The developers of `cucumber` have put this warning into their wiki: 

`Think twice before you use this feature, as whatever happens in hooks is invisible to people who only read the features. You should consider using background as a more explicit alternative if the setup should be readable by non-technical people.`

The tagged hooks are there for more general use cases than what we try acomplish here. They are more useful for some common contexts than for scenario specific actions. It's not a coincidence that the documentation about them is pretty sparse. And you will hardly find  someone making use of them broadly. 

@jsuchome 
I don't have any special opinion on the scenario outlines. The text is shorter, but the `Examples` are a bit awkward. But if we achieve the same results by using it, I have nothing against it.

aspiers commented:
> One reason is that the `delete` action is not a teardown, but rather part of the test, it belongs to the images' management features and it's actually good to have it there explicitly because everyone can see it's being tested. 

That's not an argument against my proposal, because I am also proposing to have it explicitly tested via a visible scenario.  The difference is that I am proposing that deletion be tested as a separate scenario, and this should make the reporting more concise and accurate.

> > I will repeat myself here: I would like to avoid using hooks as I only see them as helpers for scenario runners, not as place for implementation of test specific stuff.
> > 
> > > I don't understand this reasoning.  Why would Cucumber support associating hooks with tags if they were not intended for implementation of test-specific stuff?
> 
> The developers of `cucumber` have put this warning into their wiki: 
> 
> `Think twice before you use this feature, as whatever happens in hooks is invisible to people who only read the features. You should consider using background as a more explicit alternative if the setup should be readable by non-technical people.`

Yes, and if deletion is tested as a separate scenario, then it is not important to have the deletion teardown visible to non-technical people, because the same code paths are being explicitly tested in separate scenarios.

> The tagged hooks are there for more general use cases than what we try acomplish here.

Why?

> They are more useful for some common contexts than for scenario specific actions.

But this _is_ a common context - multiple scenarios sharing the same clean-up logic.

> It's not a coincidence that the documentation about them is pretty sparse. And you will hardly find  someone making use of them broadly.

It's not uncommon for documentation to be sparse in any project ;-)  Even if it's not a well-trodden path, it doesn't mean it's wrong.

> @jsuchome 
> I don't have any special opinion on the scenario outlines. The text is shorter, but the `Examples` are a bit awkward. But if we achieve the same results by using it, I have nothing against it.

I think the outline is a big improvement.  My main complaint with the existing PR is the DRY violations, and the outline really helps with that.  If the PR is amended to use an outline then I can tolerate the mixing of image creation and deletion tests within each scenario and will give a +1.

Commit:
use scenario outline to describe 3 almost identical tests

remove the tags (there's just one outline now)

wait longer for finding the image as active

vuntz commented:
@jsuchome is your last commit addressing @aspiers's comment about using an outline? (if yes, then maybe we can get some +1 love :-))

jsuchome commented:
@vuntz yes, it switched to using outlines, but now I do not have working cloud so I cannot test it. Therefore I wouldn't give +1 even to myself...

jsuchome commented:
And now I can confirm that it is working

aspiers commented:
+1, thanks for your patience with me ;-)

jsuchome commented:
I hope we can learn from this one (and from https://github.com/SUSE-Cloud/cct/pull/1) and next time we'll know better.

aspiers commented:
Sorry! But I'm changing to a +0.5, because as I said in #1, I think you should drop the Rake stuff entirely, including the scenario tags, since it just duplicates existing Cucumber functionality badly.   [Upstream recommends avoiding running Cucumber via `rake`](https://github.com/cucumber/cucumber/wiki/using-rake), and [the `-n` option already gives you the same functionality and much more](https://www.relishapp.com/cucumber/cucumber/docs/cli/run-feature-elements-matching-a-name-with-name-n).

jsuchome commented:
@aspiers Could I list features this way as easily as with simple `rake help`? And get the similar result? It doesn't seem so, your links do not seem to cover that.

And as for running tests, it seems that the name (in cucumber --name) has to include string from feature name. I like `rake feature:images` more.

vmoravec commented:
@aspiers I already explained the reasons for this decision earlier, the cost for having tags with scenarios are actually non-existent. The reason for not using rake because it `adds some additional startup time when running features` (which is 1 or 2 seconds) is not that significant when our features' runtime will be counted in minutes. Additionally, having tags with scenarios make them better accessible  with cucumber jenkins plugin; e.g. using `rake -T feature:admin` gives you list of all tests for admin node; I find that very practical and convenient compared to the plain `cucumber` command. The rake tasks are also going to be used for other purposes which will be exposed a bit later.

@jsuchome I tested the feature with virtual cloud on my local work station and the images fail to get activated within the 30 seconds period. It works fine for 120 seconds though, I tested 60, 90 and 120 several times and only the last one works for me.

aspiers commented:
@jsuchome wrote:

> @aspiers Could I list features this way as easily as with simple rake help? And get the similar result? It doesn't seem so, your links do not seem to cover that.

Yes, simply define a `rake` task which does:

```
cucumber -d | sed -n '/^Feature: /s///p'
```

or if you want the scenarios too:

```
cucumber -d | egrep '^(Feature|  Scenario):'
```

> And as for running tests, it seems that the name (in cucumber --name) has to include string from feature name.

No, that's not accurate.  ["The `--name NAME` option runs only scenarios which match a certain name. The NAME can be a substring of the names of Features, Scenarios, Scenario Outlines or Example blocks."](https://www.relishapp.com/cucumber/cucumber/docs/cli/run-feature-elements-matching-a-name-with-name-n)

Additionally, ["You can choose to run a specific scenario using the file:line format, or you can pass in a file with a list of scenarios using @-notation. The line number can fall anywhere within the body of a scenario, including steps, tags, comments, description, data tables or doc strings."](https://www.relishapp.com/cucumber/cucumber/docs/cli/run-specific-scenarios)

> I like `rake feature:images` more.

```
cucumber -d features/images.feature
```

does the same thing, and it's clearer that there's no extra magic and that it's only testing the feature in that file.

Having said that, if you want to define `feature:images` `rake` task which gives that syntactic sugar, I'm totally fine with that.  What I don't like is creating a tag and a `rake` task for every single scenario.  That just creates extra bloat and maintenance burden for no benefit.

jsuchome commented:
So far, I do not have any more scenarios here, so also no tags. 

And having current rake tasks doesn't prevent anyone from using direct calls of cucumber.

aspiers commented:
> @aspiers I already explained the reasons for this decision earlier, the cost for having tags with scenarios are actually non-existent.

That's definitely not true - if you had lots of scenarios across many feature files and each one had a unique tag and `rake` task, the cost would be extra bloat and extra maintenance.  But at least this PR does not have that problem any more :-)

> The reason for not using rake because it `adds some additional startup time when running features` (which is 1 or 2 seconds) is not that significant when our features' runtime will be counted in minutes.

I am sure there will be features in the future which only take seconds or less to run, but I agree this is not a big deal.  My main concern is not the startup time but the extra maintenance overhead required to maintain tasks for every feature we add / change / remove.  The process for contributing to the test suite should be as lightweight as possible, in order to encourage more people to contribute.  Another way of looking at it is that the `rake` tasks are really duplicating structure which is already present in the feature files, so it's technically a DRY violation.

> Additionally, having tags with scenarios make them better accessible with cucumber jenkins plugin; e.g. using `rake -T feature:admin` gives you list of all tests for admin node

But that's just duplicating existing functionality which cucumber already has - where's the added benefit over

```
cucumber -t @admin
```

?

> I find that very practical and convenient compared to the plain `cucumber` command.

Why?  It's considerably less flexible than [the comprehensive tag logic Cucumber offers](https://www.relishapp.com/cucumber/cucumber/docs/cli/tag-logic), and it's also slightly longer to type.

> The rake tasks are also going to be used for other purposes which will be exposed a bit later.

OK, if there are other reasons for needing `rake` tasks then please share them in case I can avoid arguing against doing something we genuinely need.  But so far I didn't hear any convincing reason why we need them.

aspiers commented:
> So far, I do not have any more scenarios here, so also no tags.

Right - I forgot that you changed this.

> And having current rake tasks doesn't prevent anyone from using direct calls of cucumber.

Sure, but that misses my point about DRY and maintenance burden.

aspiers commented:
I previously said:

> Having said that, if you want to define `feature:images rake` task which gives that syntactic sugar, I'm totally fine with that.

but I think I changed my mind again since thinking about DRY.  It's adding an extra layer of complexity and I can't see any benefit.  If I'm missing something then please explain.

jsuchome commented:
Well, what about this?

> Yes, simply define a rake task which does:
> cucumber -d | sed -n '/^Feature: /s///p'

for this task (simple listing of features) you are proposing to add a rake task, while we should remove ones for executing them?

aspiers commented:
> for this task (simple listing of features) you are proposing to add a rake task, while we should remove ones for executing them?

Yes, that makes sense, because listing all the features / scenarios is a simple use-case which can be covered by O(1) amount of work (one or two rake tasks), whereas adding tasks for executing them is something like O(n) because every time you add a new feature (or maybe even scenario depending on what you are proposing) you would have to add a new rake task to cover execution of that, and the extra effort spent doesn't win you anything cucumber can't already do.

jsuchome commented:
Well, we can add new tasks later if we want to.

aspiers commented:
+1, let's merge this as I'm already feeling guilty how much I delayed it ;-)
